### PR TITLE
Training Stompy

### PR DESCRIPTION
**Pull request by michael-lutz**
Added Stompy XML that makes it through the training iterations.

This updated version of Stompy has collision capsules for feet, standardized control ranges, and (slightly) reduced meshes & joints. For example, the grippers are disabled for now.

The main blocker is that Stompy begins from an awkward angle, making it difficult to actually learn to walk. Instead, it attempts to jump and ultimately falls further downward. With that being said, training is relatively stable and very repeatable.

An example training curve is included:

Screen Shot 2024-05-28 at 10 24 36 PM

### CODE REVIEW

File: ksim/mjx_gym/experiments/stompy_walk.yaml [hunk 1]
```diff
@@ -13,17 +12,17 @@ num_updates_per_batch: 8
...
 env_name: stompy
 reward_params:
   rew_forward:
     weight: 1.25
   rew_healthy:
     weight: 5.0
-    healthy_z_lower: 0.5
-    healthy_z_upper: 2.0
+    healthy_z_lower: 0.4
+    healthy_z_upper: 10.0
```
Comment (budzianowski) [inline ksim/mjx_gym/experiments/stompy_walk.yaml]: Stompy is 1.2m - why 10?

File: ksim/mjx_gym/scripts/train.slurm [hunk 2]
```diff
@@ -2,7 +2,7 @@
 #SBATCH --job-name=robotic-locomotion-training    # Job name
 #SBATCH --output=logs/training_output_%j.log      # Output file
 #SBATCH --error=logs/training_error_%j.log        # Error file
-#SBATCH --time=24:00:00                           # Maximum execution time
+#SBATCH --time=72:00:00                           # Maximum execution time
```
Comment (budzianowski) [inline ksim/mjx_gym/scripts/train.slurm:5]: Can you add explanation why 2 tasks?

File: ksim/assets/stompy/stompy.xml [hunk 3]
```diff
@@ -0,0 +1,550 @@
+<mujoco model="robot">
+	<compiler angle="radian" meshdir="meshes" />
+
+	<option timestep="0.005" iterations="20" solver="Newton" noslip_iterations="5" integrator="RK4">
```
Comment (budzianowski) [inline ksim/assets/stompy/stompy.xml:6]: Add the comment explaining at what point engine breaks with less iterations.

File: ksim/assets/stompy/stompy.xml [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+								</body>
+								<body name="link_legs_1_right_leg_1_ankle_pulley_48t_1" pos="0.24 -0.05 0" quat="0.707107 0 0 0.707107">
+									<joint name="joint_legs_1_right_leg_1_ankle_revolute" pos="0 0 0" axis="0 0 -1" range="-0.785398 0.785398" actuatorfrcrange="-6 6" />
+									<geom type="mesh" class="visual" rgba="0.615686 0.811765 0.929412 1" mesh="legs_1_right_leg_1_ankle_pulley_48t_1" />
+									<geom pos="0 -0.0270536 0.002" quat="-2.32051e-08 1 0 0" type="mesh" class="visual" rgba="0.647059 0.647059 0.647059 1" mesh="legs_1_right_leg_1_ankle_1" />
+									<geom pos="0.0645536 -0.0663036 0.00154071" quat="0.5 0.5 0.5 0.5" type="mesh" class="visual" rgba="0.972549 0.529412 0.00392157 1" mesh="legs_1_right_leg_1_x4_1_inner_1" />
+									<body name="link_legs_1_right_leg_1_x4_1_outer_1" pos="0.0990536 -0.070981 0.0120465" quat="0.707107 3.53553e-08 0.707107 1.64085e-08">
+										<joint name="joint_legs_1_right_leg_1_x4_1_dof_x4" pos="0 0 0" axis="0 0 -1" range="-0.785398 0.785398" actuatorfrcrange="-1.5 1.5" />
+
```
Comment (budzianowski) [inline ksim/assets/stompy/stompy.xml]: Remove one - you copy the object twice?

File: ksim/assets/stompy/stompy.xml [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+								<geom pos="0.0705 0.05 0.0535" quat="-2.00962e-08 -0.5 -0.866025 -1.16026e-08" type="mesh" class="visual" rgba="0.980392 0.713726 0.00392157 1" mesh="legs_1_left_leg_1_x6_1_pcb_1" />
+								<body name="link_legs_1_left_leg_1_ankle_pulley_48t_1" pos="0.24 0.05 0" quat="0.707107 0 0 -0.707107">
+									<joint name="joint_legs_1_left_leg_1_ankle_revolute" pos="0 0 0" axis="0 0 -1" range="-0.785398 0.785398" actuatorfrcrange="-6 6" />
+									<geom type="mesh" class="visual" rgba="0.615686 0.811765 0.929412 1" mesh="legs_1_left_leg_1_ankle_pulley_48t_1" />
+									<geom pos="0 0.0270536 0.021" type="mesh" class="visual" rgba="0.647059 0.647059 0.647059 1" mesh="legs_1_left_leg_1_ankle_1" />
+									<geom pos="0.0645536 0.0663036 0.0214593" quat="0.5 -0.5 0.5 -0.5" type="mesh" class="visual" rgba="0.972549 0.529412 0.00392157 1" mesh="legs_1_left_leg_1_x4_1_inner_1" />
+									<body name="link_legs_1_left_leg_1_x4_1_outer_1" pos="0.0990536 0.070981 0.0109535" quat="1.89469e-08 -0.707107 0 -0.707107">
+										<joint name="joint_legs_1_left_leg_1_x4_1_dof_x4" pos="0 0 0" axis="0 0 -1" range="-0.785398 0.785398" actuatorfrcrange="-1.5 1.5" />
+
```
Comment (budzianowski) [inline ksim/assets/stompy/stompy.xml]: Remove one - you copy the object twice?

**Review by budzianowski — COMMENTED** [review]
This is getting into a good shape! Left only some nits for now.

**Comment by codekansas** [comment]
incredible :)

File: ksim/mjx_gym/scripts/train.slurm (hunk 2 above)
Comment (michael-lutz) [reply ksim/mjx_gym/scripts/train.slurm:5]: Not sure why, but GPU's don't get fully utilized unless this value is set to 2. @codekansas do you have any ideas?

File: ksim/mjx_gym/experiments/stompy_walk.yaml (hunk 1 above)
Comment (michael-lutz) [reply ksim/mjx_gym/experiments/stompy_walk.yaml]: Ah I set this pretty high to isolate for errors, moving down

File: ksim/assets/stompy/stompy.xml [hunk 6]
```diff
@@ -0,0 +1,491 @@
+<mujoco model="robot">
+	<compiler angle="radian" meshdir="meshes" />
+	<!-- If timestep is increased, iterations reduced, etc. PPO will find a 
+	    to exploit the simulation and farm the healthy reward. Simulation
+		stability matters a lot here. -->
```
Comment (budzianowski) [inline ksim/assets/stompy/stompy.xml:5]: nit - can you add the info where is the tipping point? 20 is still a lot :(

File: ksim/mjx_gym/envs/stompy_env/stompy.py [hunk 7]
```diff
@@ -50,6 +50,7 @@ def __init__(
         self._exclude_current_positions_from_observation = exclude_current_positions_from_observation
         self._log_reward_breakdown = log_reward_breakdown
 
+        self.initial_qpos = mj_model.keyframe("default").qpos
```
Comment (budzianowski) [inline ksim/mjx_gym/envs/stompy_env/stompy.py:53]: nice

File: ksim/mjx_gym/envs/stompy_env/rewards.py [hunk 8]
```diff
@@ -74,10 +74,10 @@ def forward_reward_fn(
     Returns:
         A float wrapped in a jax array.
     """
```
Comment (budzianowski) [inline ksim/mjx_gym/envs/stompy_env/rewards.py:76]: Can you add the comment here why we switch axes?

File: README.md [hunk 9]
```diff
@@ -130,6 +130,14 @@ conda install cudnn
 
 Another common issue occurs when rendering on a headless server. For now, we recommend rendering locally or using a remote desktop connection to view the rendering.
 
+## MJCF Reference
+To visualize any MJCF file, you can run the following command:
+```bash
+python3 -m mujoco.viewer --mjcf <path-to-mjcf-file>
+```
+
+The command above loads MuJoCo's GUI, which allows you to simulate the model, manually specify joints, and
```
Comment (budzianowski) [inline README.md:139]: Finish the sentence?

File: README.md [hunk 10]
```diff
@@ -130,6 +130,14 @@ conda install cudnn
...
+## MJCF Reference
+To visualize any MJCF file, you can run the following command:
+```bash
+python3 -m mujoco.viewer --mjcf <path-to-mjcf-file>
+```
+
+The command above loads MuJoCo's GUI, which allows you to simulate the model, manually specify joints, and 
+
 ## TODO
 - [ ] Get Stompy to load efficiently with MJX (currently, the meshes and collision detection are not loading correctly)
```
Comment (budzianowski) [inline README.md:142]: Done

**Review by budzianowski — APPROVED** [review]
A couple of small nits but otherwise lgtm! Can you also share a video of the current 'best' policy.

**Pull request merged**